### DISCUSSION TITLE
BMS Aggregate to CAN BUS

### DISCUSSION THREAD

**Discussion by RoySalisbury** [syssi/esphome-jk-bms#233] [Ideas]
I see that there are multiple different esphome projects here and it got me thinking ...

I wonder if its possible to create an "aggregate" device that reads in multiple BMS's (say 6 JK BMS devices over BLE/TCP/UART) and outputs them as a single device over CAN ... or as multiple devices over CAN like you daisy chained them together.

Most of the "commercial" type BMS systems (Seplos, Signature Solar, SOK, ect) will daisy chain all the individual BMS boxes together so that you can plug them into a single inverter.  However, the DIY versions (JK, JBD, ect) are all stand alone with no way to connect them together as a single communications line.

In simple terms, each device would have its own address on the bus, and the "master" device would just do the translation to each BMS over its own dedicated connection (BLE/TCP/UART).  

You could almost let HomeAssistant be the "brains" and the ESP devices attached via CAN just talk via the HomeAssistant API to get the necessary info.

This would actually be something that I (and probably others) would pay for.  No ones has this.

**Comment by harald7** [comment]
Hi Roy

What system you are running?

**Comment by RoySalisbury** [comment]
Are you referring to OS or Inverter/CC/BMS?

The OS can be any of the following (MacOS, Linuux, Windows).  For the others

EG4 6500EX inverter/cc
JK BMS 

If thats not what your looking for, please elaborate.

Roy

**Reply by harald7** [reply]
I refer to the solar system.
I got 5 battery packs with jk in are victron system.

I calculate in iobroker the charge & discharge current from the esp32 data for each pack, accordingly to that rules:
SOC (State Of Charge) from the BMS
•	CCCM_SOC_ENABLE = True/False
•	DCCM_SOC_ENABLE = True/False
CCCM limits the charge/discharge current depending on the SOC
o	between 99% - 100% => 5A charge
o	between 95% - 98% => 1/4 Max charge
o	between 91% - 95% => 1/2 Max charge
o	30% - 91% => Max charge and Max discharge
o	between 20% - 30% => 1/2 Max discharge
o	between 10% - 20% => 1/4 Max discharge
o	below <= 10% => 5A
Cell Voltage
•	CCCM_CV_ENABLE = True/False
•	DCCM_CV_ENABLE = True/False
CCCM limits the charge/discharge current depending on the highest/lowest cell voltages
o	between 3.50V - 3.55V => 2A charge
o	between 3.45V - 3.50V => 30A charge
o	between 3.30V - 3.45V => 60A
o	3.30V - 3.10V => Max charge and Max discharge (60A)
o	between 2.90V - 3.10V => 30A discharge
o	between 2.8V - 2.9V => 5A discharge
o	below <= 2.70V => 0A discharge
 
The sum of charge & discharge current from each pack is sended via modbus to the victron system.
I think that's the maximum what communication between battery and solar system can do.

Maybe it's possible to send the data from iobroker via can bus to your system.

Harald

**Comment by RoySalisbury** [comment]
After I had all my components in place and started working on the monitoring I realized that I should have went with a Victron based system.  You get what you pay for I guess.

However, I think it may be possible to do something like what you have programmatically.  The inverter does have an serial connection that the parameters can be changed over (not the CAN bus).  I may look to see if I can use Home Assistants monitoring/events to do what the Victron CCCM does.

Roy

**Comment by harald7** [comment]
If you search you will find in every soup a hair.

From my point of view the entire communication between bms and solar system comes down to 2 Parameter.
Max charge current and max discharge current that's all.

**Comment by webgress** [comment]
@RoySalisbury have you ever succeeded in setting up this communication with the inverter?
I have EG4 18K and several JK BMSs 
And I would like to be able to report charge level and max currents to inverter from ESP as if it is a "smart battery"

thank you
